bn_mp_radix_size_overestimate.c

```diff
@@ -0,0 +1,258 @@
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   mp_int bi_k_bis;
+#endif
+   int rem, bit_count;
+   mp_err e = MP_OKAY;
+
+   if ((base < 2) || (base > 64)) {
+      *size = 0;
+      return MP_VAL;
+   }
+   /* floor( log_2(a) ) + 1 */
+   bit_count = mp_count_bits(a);
+
+   if (bit_count == 0) {
+      *size = 2;
+      return MP_OKAY;
+   }
+   if (base == 2) {
+      *size = bit_count + 2;
+      return MP_OKAY;
+   }
+
+   /* A small shortcut for powers of two. */
+   if (!(base&(base-1))) {
+      unsigned int x = (unsigned int)base;
+      int y;
+      for (y=0; (y < 7) && !(x & 1u); y++) {
+         x >>= 1;
+      }
+      *size = bit_count/y;
+      rem = bit_count - ((*size) * y);
+      /* Add 1 for the remainder if any and 1 for "\0". */
+      *size += (rem == 0) ? 1 : 2;
+      /* And one extra character for the minus sign */
+      if (a->sign == MP_NEG) {
+         (*size)++;
+      }
+      return MP_OKAY;
+   }
+
+   /* d(bitcount,base) = floor( (bitcount * 2^p) / k(base) ) + 1 */
+
+   if ((e = mp_init_multi(&bi_bit_count, &bi_k, NULL)) != MP_OKAY) {
+      *size = 0;
+      return e;
+   }
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   if ((e = mp_init(&bi_k_bis)) != MP_OKAY) {
+      /* No "goto" to avoid cluttering this code with even more preprocessor branches */
+      mp_clear_multi(&bi_bit_count, &bi_k, NULL);
+      *size = 0;
+      return e;
+   }
+#endif
+   /* "long" is at least as large as "int" according to even the oldest C standards */
+   mp_set_l(&bi_bit_count, bit_count);
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   /* There is no mp_set_u16 for MP_8BIT */
+   mp_set_u32(&bi_k, logbases_high[base]);
+   if ((e = mp_mul_2d(&bi_k, LTM_RADIX_SIZE_HALF_TABLE_SCALE, &bi_k)) != MP_OKAY) {
+      *size = 0;
+      goto LTM_E1;
+   }
+   mp_set_u32(&bi_k_bis, logbases_low[base]);
+   if ((e = mp_add(&bi_k_bis, &bi_k, &bi_k)) != MP_OKAY) {
+      *size = 0;
+      goto LTM_E1;
+   }
+#else
+   mp_set_u64(&bi_k, logbases[base]);
+#endif
+   if ((e = mp_mul_2d(&bi_bit_count, LTM_RADIX_SIZE_SCALE, &bi_bit_count)) != MP_OKAY) {
+      *size = 0;
+      goto LTM_E1;
+   }
+   if ((e = mp_div(&bi_bit_count, &bi_k, &bi_bit_count, NULL)) != MP_OKAY) {
+      *size = 0;
+      goto LTM_E1;
+   }
+   /*
+      The first addition of one is done because of the truncating last division,
+      and the second addition of one is done to make room for NIL ('\0').
+
+      Casting from "long" to "int" can be done because "bi_bit_count" fits into an "int"
+      by definition.
+    */
+   *size = (int)mp_get_l(&bi_bit_count) + 1 + 1 ;
+
+#if ( (defined MP_8BIT) && (INT_MAX > 0xFFFF))
+   /* TODO: Add a third table? But how likely is it that "int" is 32-bit in
+            an 8-bit environment? */
+   /* diff. is 3 bits, hence add 2 (two) */
+   *size += 2;
+#endif
+
+   /* And before we forget it: one extra character for the minus sign */
+   if (a->sign == MP_NEG) {
+      (*size)++;
```

Comment:
Why not always add 1 for the sign? We are overestimating?

Reply:
> Why not always add 1 for the sign? We are overestimating

Overestimating by at most 1 (one) like the GMP function (with the exception for MP_8BIT where it is 3 at most). That was my goal, it had quite some work involved to reach it and I won't move any posts here.

Reply:
Why not. As I said before I would rather like a simpler function with a less precise estimate.

Reply:
> As I said before I would rather like a simpler/more compact function with a less precise estimate.

By saving a single branch, probably less than half a dozen CPU operations?
Really?

Reply:
It is more a question of intend. I would write it differently, but it doesn't really matter. It should reduce to a single add operation.

*size += sign;

Or

*size += sign ? 1 : 0;

Reply:
Trying to be GMP compatible makes somehow sense, but also I wouldn't care if we'd choose to be sometimes off-by-two or even off-by-X if it would make the implementation a lot easier.

@minad Doesn't this boil down to only a style question here? Or what do you see as an advantage of writing `*size += sign ? 1 : 0;` over the way it is now?

Reply:
No advantage. I would write *size += 1 :)

Reply:
> if it would make the implementation a lot easier.

```text
$ gcc --version
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
$ gcc -g3 -O3 -c bn_mp_radix_size_overestimate.c
$ objdump -d -S bn_mp_radix_size_overestimate.o
...
   if (a->sign == MP_NEG) {
 1e6:	83 7d 08 01          	cmpl   $0x1,0x8(%rbp)
 1ea:	75 91                	jne    17d <mp_radix_size_overestimate+0x17d>
      (*size)++;
 1ec:	83 c0 03             	add    $0x3,%eax
 1ef:	41 89 04 24          	mov    %eax,(%r12)
 1f3:	eb 88                	jmp    17d <mp_radix_size_overestimate+0x17d>
 1f5:	0f 1f 00             	nopl   (%rax)
...
```

If you want to set `*size = 1` at the start instead you save about 5 CPU-cycles (3 if you compile with `-Os) with my AMD A8-6600K.
In the pale face of the imminent danger to repeat myself: really?
But your definition of "a **lot** easier" [emphasis by me] might differ ;-)

> Doesn't this boil down to only a style question here?

Yes, and that means: I don't care. Just tell me how you want it and I'll do it and if I feel especially adventurous at that time I might even add a note to the wiki-page ;-)

Reply:
This is not about optimization but about simplicity. Why write something more complicated than necessary?